File: strings/prefix_and_suffix_search.rb

class WordFilter

=begin
    :type words: String[]
=end
    def initialize(words)
        @dict = {}

      words.each_with_index do |word, i|
        len = word.size
        
        prefixes = Array.new(len + 1) { |j| word[0, j]}
        suffixes = Array.new(len + 1) { |j| word[len - j, j] }
        
        prefixes.each do |prefix|
          suffixes.each do |suffix|
            @dict["#{prefix}##{suffix}"] = i
          end
        end
      end
    end

=begin
    :type pref: String
    :type suff: String
    :rtype: Integer
=end
    def f(pref, suff)
        @dict["#{prefix}##{suffix}"] || -1
    end


end

# Your WordFilter object will be instantiated and called as such:
# obj = WordFilter.new(words)
# param_1 = obj.f(pref, suff)

File: strings/prefix_and_suffix_search.py
class WordFilter:
    from collections import defaultdict
    def __init__(self, words: List[str]):
        self.dct = {}
        for i in range(len(words)):
            for j in range(len(words[i])):
                for k in range(len(words[i])-1, -1, -1):
                    self.dct[(words[i][:j + 1], words[i][k:])] = i

    def f(self, prefix: str, suffix: str) -> int:
        return self.dct.get((prefix, suffix), -1)


# Your WordFilter object will be instantiated and called as such:
# param_1 = obj.f(pref,suff)
# obj = WordFilter(words)